apps/backend/app/core/algorithms.py

```
@@ -0,0 +1,479 @@
+from scipy.sparse import csr_matrix, dok_matrix
+from functools import lru_cache
+import asyncio
+from concurrent.futures import ThreadPoolExecutor
+
+# Pre-compiled regex patterns for performance
+WORD_PATTERN = re.compile(r'\b\w+\b')
+SENTENCE_PATTERN = re.compile(r'[.!?]+')
+STOP_WORDS = frozenset({
+    'a', 'an', 'and', 'are', 'as', 'at', 'be', 'by', 'for', 'from',
+    'has', 'he', 'in', 'is', 'it', 'its', 'of', 'on', 'that', 'the',
+    'to', 'was', 'will', 'with', 'the', 'this', 'but', 'they', 'have',
```

Ruff (0.11.9)

Sets should not contain duplicate item `'the'`

Remove duplicate item

(B033)